packages/strapi-admin/services/permission/sections-builder/handlers.js

```diff
@@ -0,0 +1,134 @@
+'use strict';
+
+const { curry, matchesProperty, pick } = require('lodash/fp');
+
+const utils = {
+  isOfKind: kind => matchesProperty('kind', kind),
+  resolveContentType: uid => strapi.contentTypes[uid],
+  isNotInSubjects: subjects => uid => !subjects.find(subject => subject.uid === uid),
+  hasProperty: curry((property, subject) => {
+    return !!subject.properties.find(prop => prop.value === property);
+  }),
+  getValidOptions: pick(['applyToProperties']),
+  toSubjectTemplate: ct => ({ uid: ct.uid, label: ct.info.name, properties: [] }),
+};
+
+/**
+ * Transforms & adds the given  setting action to the section
+ * Note: The action is transformed to a setting specific format
+ */
+const settings = ({ action, section }) => {
+  const { category, subCategory, displayName, actionId } = action;
+
+  section.push({
+    displayName,
+    category,
+    subCategory,
+    action: actionId,
+  });
+};
+
+/**
+ * Transforms & adds the given plugin action to the section
+ * Note: The action is transformed to a plugin specific format
+ */
+const plugins = ({ action, section }) => {
+  const { pluginName, subCategory, displayName, actionId } = action;
+
+  section.push({
+    displayName,
+    plugin: pluginName,
+    subCategory,
+    action: actionId,
+  });
+};
+
+/**
+ * Transforms & adds the given action to the section's actions field
+ * Note: The action is transformed to a content-type specific format
+ */
+const contentTypesBase = ({ action, section }) => {
+  const { displayName, actionId, subjects, options } = action;
+
+  section.actions.push({
```

Comment:
addAction instead of leaking internal code of the section ?

Reply:
Since we don't have the same shape for every section (plugins/settings !== ST/CT), we can't have a common domain API for the section and we need to trust each handler :confused: